pygsp/graphs/nngraphs/nngraph.py

```diff
@@ -1,199 +1,560 @@
+    ...             features = rs.uniform(size=(size, dim))
+    ...             start = time.time()
+    ...             _ = graphs.NNGraph(features, backend=backend)
+    ...             times[i][j][k] = time.time() - start
+    >>> fig, axes = plt.subplots(1, 2, sharey=True)
+    >>> for j, (dim, ax) in enumerate(zip(dims, axes)):
+    ...     for k, backend in enumerate(backends):
+    ...         _ = ax.loglog(sizes, times[:, j, k], '.-', label=backend)
+    ...         _ = ax.set_title('{} dimensions'.format(dim))
+    ...         _ = ax.set_xlabel('number of vertices')
+    >>> _ = axes[0].set_ylabel('execution time [s]')
+    >>> _ = axes[1].legend(loc='upper left')
 
-        W = sparse.csc_matrix((spv, (spi, spj)), shape=(N, N))
+    """
 
-        # Sanity check
-        if np.shape(W)[0] != np.shape(W)[1]:
-            raise ValueError('Weight matrix W is not square')
+    def __init__(self, features, standardize=False,
+                 metric='euclidean', order=3,
+                 kind='knn', k=10, radius='estimate-knn',
+                 kernel='gaussian', kernel_width=None,
+                 backend='scipy-ckdtree',
+                 **kwargs):
+
+        # features is stored in coords, potentially standardized
+        self.standardize = standardize
+        self.metric = metric
+        self.kind = kind
+        self.kernel = kernel
+        self.k = k
+        self.backend = backend
+
+        features = np.asanyarray(features)
+        if features.ndim != 2:
+            raise ValueError('features should be #vertices x dimensionality')
+        n_vertices, dimensionality = features.shape
+
+        params_graph = dict()
+        for key in ['lap_type', 'plotting']:
+            try:
+                params_graph[key] = kwargs.pop(key)
+            except KeyError:
+                pass
+
+        if standardize:
+            # Don't alter the original data (users would be surprised).
+            features = features - np.mean(features, axis=0)
+            features /= np.std(features, axis=0)
+
+        # Order consistent with metric (used by kdtree and ckdtree).
+        _orders = {
+            'euclidean': 2,
+            'manhattan': 1,
+            'max_dist': np.inf,
+            'minkowski': order,
+        }
+        order = _orders.pop(metric, None)
+
+        if kind == 'knn':
+            if not 1 <= k < n_vertices:
+                raise ValueError('The number of neighbors (k={}) must be '
+                                 'greater than 0 and smaller than the number '
+                                 'of vertices ({}).'.format(k, n_vertices))
+            radius = None
+        elif kind == 'radius':
+            if radius == 'estimate':
+                maximums = np.amax(features, axis=0)
+                minimums = np.amin(features, axis=0)
+                distance_max = np.linalg.norm(maximums - minimums, order)
+                radius = distance_max / np.power(n_vertices, 1/dimensionality)
```

Comment:
So you removed the max(, 3). I guess this will fail completly with many datasets. For example everything in computer vision (just try MNSIT...)

Reply:
Furhermore, you probably need a constant that multiply the radius (i.e the average number of points in the disk), because now it is probably `2`. I think it should be
``` radius = distance_max / np.power(5*n_vertices, 1/dimensionality)```
for `10` points on the disk, but I am not sure.
The figure you send me, you could just compute additionally the number of neighbors in a disk... And find the right constant. This constant is going to depend on the distribution. I would just assume uniform distribution...

Reply:
Thanks for raising this, it's an important discussion to have. Here's what was before for reference:
```
bounding_radius = 0.5 * np.linalg.norm(np.amax(Xout, axis=0) -
                                       np.amin(Xout, axis=0), 2)
scale = np.power(N, 1. / float(min(d, 3))) / 10.
Xout *= scale / bounding_radius
```
* So `min(d, 3)` was removed. Why upper-bound the data dimensionality to 3?
* Is the `/ 10` the constant you're talking about, to have 10 points in the ball?